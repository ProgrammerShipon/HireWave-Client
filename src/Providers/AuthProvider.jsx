import { GithubAuthProvider, GoogleAuthProvider, createUserWithEmailAndPassword, getAuth, onAuthStateChanged, signInWithEmailAndPassword, signInWithPopup, signOut, updatePassword, updateProfile } from "firebase/auth";
import { createContext, useEffect, useState } from 'react';
import useAxiosSecure from "../Hooks/useAxiosSecure";
import app from '../firebase/firebase.config';

export const AuthContext = createContext();

const AuthProvider = ({ children }) => {
    const [axiosSecure] = useAxiosSecure();
    const [user, setUser] = useState({});
    const [currentUser, setCurrentUser] = useState({});
    const [loading, setLoading] = useState(true);

    // auth initialize
    const auth = getAuth(app);

    // sign up user
    const signUpUser = (email, password) => {
        setLoading(true);
        return createUserWithEmailAndPassword(auth, email, password);
    };

    // email sign in
    const signIn = (email, password) => {
        setLoading(true);
        return signInWithEmailAndPassword(auth, email, password);
    };

    // update profile
    const profileUpdate = (currentUser, name, photoLink) => {
        setLoading(true);
        return updateProfile(currentUser, {
            displayName: name,
            photoURL: photoLink,
        });
    };

    // change password
    const changePassword = (newPassword) => {
        setLoading(true)
        const user = auth.currentUser;
        return updatePassword(user, newPassword);
    }

    // google sign in
    const googleSignIn = () => {
        setLoading(true);
        const googleProvider = new GoogleAuthProvider();
        return signInWithPopup(auth, googleProvider)
    }

    // github sign in
    const gitHubSignIn = () => {
        setLoading(true);
        const gitHubProvider = new GithubAuthProvider();
        return signInWithPopup(auth, gitHubProvider);
    };

    // log out
    const logOut = () => {
        return signOut(auth);
    };

    useEffect(() => {
        setLoading(true);
        const unsubscribe = onAuthStateChanged(auth, (authUser) => {
            setUser(authUser)
            if (authUser?.email) {
                axiosSecure(`/users/email/${authUser?.email}`)
                    .then((data) => {
                        setCurrentUser(data.data);
                        setLoading(false);
                    })
                    .catch((err) => {
                        console.log(err)
                        setLoading(false);
                    });
            }
            if (authUser === null) {
                setLoading(false);
            }

            // if (authUser) {
            //     axiosSecure
            //         .post("/jwt", { email: authUser.email })
            //         .then((response) => {
            //             localStorage.setItem("access-token", response.data.token);
            //             setLoading(false);
            //         })
            //         .catch((error) => {
            //             console.log(error);
            //         });
            // }
            // localStorage.removeItem("access-token");
        });

        return () => {
            return unsubscribe();
        }
    }, []);


    const authInfo = {
        user,
        loading,
        currentUser,
        signUpUser,
        signIn,
        profileUpdate,
        googleSignIn,
        gitHubSignIn,
        logOut,
        changePassword
    }

    return (
        <AuthContext.Provider value={authInfo}>
            {children}
        </AuthContext.Provider>
    );
};

export default AuthProvider;
