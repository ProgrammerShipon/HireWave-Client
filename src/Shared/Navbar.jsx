import React, { useEffect, useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import { FaBars, FaSearch } from "react-icons/fa";
import Logo from '../Assets/images/logo-01.png';
import NavItems from "../Components/NavItems";
import SideNav from "./SideNav";
import { toast } from "react-toastify";
import useAllJobs from "../Hooks/useAllJobs";

const Navbar = () => {
    const [isOpen, setIsOpen] = useState(false);
    const [navState, setNavState] = useState(false);
    const [allJobsData, loading] = useAllJobs();
    const [filteredData, setFilteredData] = useState([]);
    const [searchTerm, setSearchTerm] = useState('');
    const navigate = useNavigate();

    const onNavScroll = () => {
        if (typeof window !== 'undefined' && window.scrollY > 300) {
            setNavState(true);
        } else {
            setNavState(false);
        }
    };

    useEffect(() => {
        if (typeof window !== 'undefined') {
            window.addEventListener("scroll", onNavScroll);
        }

        return () => {
            if (typeof window !== 'undefined') {
                window.removeEventListener("scroll", onNavScroll);
            }
        };
    }, []);

    useEffect(() => {
        if (!loading) { // Add this condition to check if allJobsData is defined
            const searchTitle = searchTerm ? searchTerm.toLowerCase() : "";
            const filter = allJobsData.filter((job) =>
                (!searchTitle || job.title?.toLowerCase().includes(searchTitle) || job.companyName?.toLowerCase().includes(searchTitle))
            );

            setFilteredData(filter);
        }
    }, [searchTerm, allJobsData.length]);

    const handleSearch = () => {
        if (searchTerm === '') {
            toast.warn('Write search term!', {
                position: "top-right",
                autoClose: 3000,
                hideProgressBar: false,
                closeOnClick: true,
                pauseOnHover: true,
                draggable: true,
                theme: "light",
            });
        } else {
            navigate('/search_results', { state: { filteredData } });
            setSearchTerm('');
        }
    };

    const toggle = () => {
        setIsOpen(!isOpen);
    };

    return (
        <header className={`fixed w-full shadow-xl duration-300 ease-in-out ${navState ? 'backdrop-blur-md bg-white/60 py-2 shadow-green/10' : 'bg-white py-3 backdrop-blur-md'} z-50`}>
            <nav className="container flex items-center justify-between">
                {/* logo */}
                <Link to="/">
                    <img data-testid="logo" className="w-52" src={Logo} alt="hire wave logo" />
                </Link>

                <div className="flex items-center gap-8">
                    {/* search bar */}
                    <div className={`hidden w-full md:w-52 lg:w-72 lg:flex items-center bg-white rounded-md border overflow-hidden duration-300 ${navState ? 'scale-x-100 opacity-100' : 'scale-x-0 opacity-0'}`}>
                        <input
                            data-testid="search-input"
                            type="text"
                            placeholder="Job Title / Keywords or Company"
                            className="w-full py-2 pl-3 rounded-s-md bg-transparent focus:outline-none"
                            value={searchTerm}
                            onChange={(e) => setSearchTerm(e.target.value)}
                        />
                        <button onClick={handleSearch} className="bg-dark text-white px-4 py-3 hover:text-green duration-300">
                            <FaSearch />
                        </button>
                    </div>

                    {/* nav items */}
                    <ul className="hidden lg:flex items-center gap-8">
                        <NavItems />
                    </ul>
                </div>

                {/* toggle button */}
                <button
                    data-testid="toggle__button"
                    onClick={toggle}
                    className="lg:hidden text-dark">
                    <FaBars className="text-3xl" />
                </button>

                {/* side navbar */}
                <SideNav isOpen={isOpen} toggle={toggle} searchTerm={searchTerm} setSearchTerm={setSearchTerm} handleSearch={handleSearch} />
            </nav>
        </header>
    );
};

export default Navbar;