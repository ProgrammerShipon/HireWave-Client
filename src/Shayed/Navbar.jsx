import { Link, NavLink } from "react-router-dom";

const Navbar = () => {
  return (
    <div>
      <nav>
        <div>
          <Link to="/"> HireWave</Link>
        </div>

        <ul>
          <li>
            <NavLink to="/"> Home </NavLink>
          </li>
        </ul>

        <ul>
          <li>
            <NavLink to="/about"> About </NavLink>
          </li>
        </ul>
      </nav>
    </div>
  );
};

export default Navbar;
