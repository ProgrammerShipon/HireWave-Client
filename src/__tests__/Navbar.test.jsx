import TestRenderer from 'react-test-renderer';
import { MemoryRouter } from "react-router";
import { QueryClient, QueryClientProvider, } from '@tanstack/react-query';
import Navbar from "../Shared/Navbar";
import AuthProvider from "../Providers/AuthProvider";
const queryClient = new QueryClient();

describe('Navbar.jsx', () => {
    it("renders correctly", () => {
        const tree = TestRenderer.create(
            <MemoryRouter>
                <QueryClientProvider client={queryClient}>
                    <AuthProvider>
                        <Navbar />
                    </AuthProvider>
                </QueryClientProvider>
            </MemoryRouter>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("renders the logo", () => {
        const component = TestRenderer.create(
            <MemoryRouter>
                <QueryClientProvider client={queryClient}>
                    <AuthProvider>
                        <Navbar />
                    </AuthProvider>
                </QueryClientProvider>
            </MemoryRouter>
        );
        const logo = component.root.findByProps({
            "data-testid": "logo",
        });
        expect(logo.props.alt).toBe("hire wave logo");
    });

    it("renders navigation items", () => {
        const component = TestRenderer.create(
            <MemoryRouter>
                <QueryClientProvider client={queryClient}>
                    <AuthProvider>
                        <Navbar />
                    </AuthProvider>
                </QueryClientProvider>
            </MemoryRouter>
        );
        const navItems = component.root.findAllByType("li");
        expect(navItems.length).toBe(10);
    });

    it("renders the toggle button", () => {
        const component = TestRenderer.create(
            <MemoryRouter>
                <QueryClientProvider client={queryClient}>
                    <AuthProvider>
                        <Navbar />
                    </AuthProvider>
                </QueryClientProvider>
            </MemoryRouter>
        );
        const toggleButton = component.root.findByProps({ "data-testid": "toggle__button" });
        expect(toggleButton.props.onClick).toBeDefined();
    });
});