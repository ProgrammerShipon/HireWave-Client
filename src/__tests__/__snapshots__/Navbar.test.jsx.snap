// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar.jsx > renders correctly 1`] = `
<header
  className="fixed w-full shadow-xl duration-300 ease-in-out bg-white py-3 backdrop-blur-md z-50"
>
  <nav
    className="container flex items-center justify-between"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <img
        alt="hire wave logo"
        className="w-52"
        data-testid="logo"
        src="/src/Assets/images/logo-01.png"
      />
    </a>
    <div
      className="flex items-center gap-8"
    >
      <div
        className="hidden w-full md:w-52 lg:w-72 lg:flex items-center bg-white rounded-md border overflow-hidden duration-300 scale-x-0 opacity-0"
      >
        <input
          className="w-full py-2 pl-3 rounded-s-md bg-transparent focus:outline-none"
          data-testid="search-input"
          onChange={[Function]}
          placeholder="Job Title / Keywords or Company"
          type="text"
          value=""
        />
        <button
          className="bg-dark text-white px-4 py-3 hover:text-green duration-300"
          onClick={[Function]}
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            />
          </svg>
        </button>
      </div>
      <ul
        className="hidden lg:flex items-center gap-8"
      >
        <li>
          <a
            aria-current="page"
            className="text-green text-lg py-2 inline-block"
            href="/"
            onClick={[Function]}
          >
            Home
          </a>
        </li>
        <li>
          <a
            className="text-dark text-lg hover:text-green duration-300 py-2 inline-block"
            href="/browse_jobs"
            onClick={[Function]}
          >
            Browse Jobs
          </a>
        </li>
        <li>
          <a
            className="text-dark text-lg hover:text-green duration-300 py-2 inline-block"
            href="/candidates"
            onClick={[Function]}
          >
            Candidates
          </a>
        </li>
        <li>
          <a
            className="text-dark text-lg hover:text-green duration-300 py-2 inline-block"
            href="/recruiters"
            onClick={[Function]}
          >
            Recruiters
          </a>
        </li>
        <li>
          <a
            className="text-dark text-lg hover:text-green duration-300 py-2 inline-block"
            href="/contact"
            onClick={[Function]}
          >
            Contact
          </a>
        </li>
        <a
          className="flex items-center gap-2 py-2 text-green font-medium"
          href="/login"
          onClick={[Function]}
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 256 256"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M230.92,212c-15.23-26.33-38.7-45.21-66.09-54.16a72,72,0,1,0-73.66,0C63.78,166.78,40.31,185.66,25.08,212a8,8,0,1,0,13.85,8c18.84-32.56,52.14-52,89.07-52s70.23,19.44,89.07,52a8,8,0,1,0,13.85-8ZM72,96a56,56,0,1,1,56,56A56.06,56.06,0,0,1,72,96Z"
            />
          </svg>
          Login
        </a>
      </ul>
    </div>
    <button
      className="lg:hidden text-dark"
      data-testid="toggle__button"
      onClick={[Function]}
    >
      <svg
        className="text-3xl"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
        />
      </svg>
    </button>
    <aside
      className="lg:hidden fixed top-0 bg-white w-[280px] md:w-[350px] h-screen p-5 z-50 duration-300 ease-in -right-full"
    >
      <div
        className="flex items-center justify-between border-b border-green/20 pb-2"
      >
        <a
          href="/"
          onClick={[Function]}
        >
          <img
            alt="hire wave logo"
            className="w-40 md:w-52"
            src="/src/Assets/images/logo-01.png"
          />
        </a>
        <button
          className="p-3 bg-dark text-white rounded-full"
          onClick={[Function]}
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 384 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
            />
          </svg>
        </button>
      </div>
      <ul
        className="flex flex-col px-4 divide-y divide-green/20 mt-4"
      >
        <li>
          <a
            aria-current="page"
            className="text-green text-lg py-2 inline-block"
            href="/"
            onClick={[Function]}
          >
            Home
          </a>
        </li>
        <li>
          <a
            className="text-dark text-lg hover:text-green duration-300 py-2 inline-block"
            href="/browse_jobs"
            onClick={[Function]}
          >
            Browse Jobs
          </a>
        </li>
        <li>
          <a
            className="text-dark text-lg hover:text-green duration-300 py-2 inline-block"
            href="/candidates"
            onClick={[Function]}
          >
            Candidates
          </a>
        </li>
        <li>
          <a
            className="text-dark text-lg hover:text-green duration-300 py-2 inline-block"
            href="/recruiters"
            onClick={[Function]}
          >
            Recruiters
          </a>
        </li>
        <li>
          <a
            className="text-dark text-lg hover:text-green duration-300 py-2 inline-block"
            href="/contact"
            onClick={[Function]}
          >
            Contact
          </a>
        </li>
        <a
          className="flex items-center gap-2 py-2 text-green font-medium"
          href="/login"
          onClick={[Function]}
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 256 256"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M230.92,212c-15.23-26.33-38.7-45.21-66.09-54.16a72,72,0,1,0-73.66,0C63.78,166.78,40.31,185.66,25.08,212a8,8,0,1,0,13.85,8c18.84-32.56,52.14-52,89.07-52s70.23,19.44,89.07,52a8,8,0,1,0,13.85-8ZM72,96a56,56,0,1,1,56,56A56.06,56.06,0,0,1,72,96Z"
            />
          </svg>
          Login
        </a>
      </ul>
      <div
        className="w-full flex items-center bg-white rounded-md border overflow-hidden mt-6"
      >
        <input
          className="w-full py-2 pl-3 rounded-s-md bg-transparent focus:outline-none"
          onChange={[Function]}
          placeholder="Job Title / Keywords or Company"
          type="text"
          value=""
        />
        <button
          className="bg-dark text-white px-4 py-3"
          onClick={[Function]}
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            />
          </svg>
        </button>
      </div>
    </aside>
    <div
      className="fixed lg:hidden w-full h-screen top-0 left-0 z-30 bg-dark bg-opacity-30 duration-300 ease-in opacity-0 invisible"
      onClick={[Function]}
    />
  </nav>
</header>
`;
